File: src/main/java/org/example/algo/game2048/Game2048.java
package org.example.algo.game2048;

/**
 * @author yoyocraft
 * @date 2024/09/14
 */
public class Game2048 {

    public static void main(String[] args) {
        int[][] matrix = {
                {0, 2, 2, 0},
                {0, 0, 2, 2},
                {2, 4, 4, 2},
                {2, 4, 4, 4},
        };

        printMatrix(matrix);
        System.out.println("----------");
        String direction = "up";
        DirectionEnum directionEnum = DirectionEnum.resolve(direction);
        move(matrix, directionEnum);
        printMatrix(matrix);
    }

    public static void printMatrix(int[][] matrix) {
        for (int[] row : matrix) {
            for (int num : row) {
                System.out.print(num + " ");
            }
            System.out.println();
        }
    }

    public static void move(int[][] matrix, DirectionEnum direction) {
        int rows = matrix.length;
        int cols = matrix[0].length;

        int[][] next = getNextDirection(direction);
        switch (direction) {
            case UP:
                for (int j = 0; j < cols; j++) {
                    calculate(matrix, 0, j, next);
                }
                break;
            case DOWN:
                for (int j = cols - 1; j >= 0; j--) {
                    calculate(matrix, rows - 1, j, next);
                }
                break;
            case LEFT:
                for (int i = 0; i < rows; i++) {
                    calculate(matrix, i, 0, next);
                }
                break;
            case RIGHT:
                for (int i = rows - 1; i >= 0; i--) {
                    calculate(matrix, i, cols - 1, next);
                }
                break;
            default:
                throw new IllegalArgumentException("Invalid direction: " + direction);
        }
    }

    private static boolean inRange(int i, int j, int rows, int cols) {
        return i >= 0 && i < rows && j >= 0 && j < cols;
    }

    private static int[][] getNextDirection(DirectionEnum direction) {
        switch (direction) {
            case UP:
                return new int[][]{{1, 0}};
            case DOWN:
                return new int[][]{{-1, 0}};
            case LEFT:
                return new int[][]{{0, 1}};
            case RIGHT:
                return new int[][]{{0, -1}};
            default:
                throw new IllegalArgumentException("Invalid direction: " + direction);
        }
    }

    private static int[] nextNonZeroValue(int[][] matrix, int i, int j, int[][] next) {
        int rows = matrix.length;
        int cols = matrix[0].length;

        if (!inRange(i, j, rows, cols)) {
            return null;
        }

        int nextI = i + next[0][0];
        int nextJ = j + next[0][1];
        while (inRange(nextI, nextJ, rows, cols)) {
            if (matrix[nextI][nextJ] != 0) {
                return new int[]{nextI, nextJ, matrix[nextI][nextJ]};
            }
            nextI += next[0][0];
            nextJ += next[0][1];
        }
        return null;
    }

    private static void calculate(int[][] matrix, int i, int j, int[][] next) {
        int rows = matrix.length;
        int cols = matrix[0].length;
        if (!inRange(i, j, rows, cols)) {
            return;
        }

        int[] result = nextNonZeroValue(matrix, i, j, next);
        if (result == null) {
            return;
        }

        int nextI = result[0];
        int nextJ = result[1];
        int nextValue = result[2];

        if (matrix[i][j] == 0) {
            matrix[i][j] = nextValue;
            matrix[nextI][nextJ] = 0;
            calculate(matrix, i, j, next);
        } else if (matrix[i][j] == nextValue) {
            matrix[i][j] *= 2;
            matrix[nextI][nextJ] = 0;
        }

        int newI = i + next[0][0];
        int newJ = j + next[0][1];
        calculate(matrix, newI, newJ, next);
    }
}

File: src/main/java/org/example/algo/game2048/game2048.js
Array.prototype.print = function () {
    console.log(this.join('\n'));
}

const matrix = [
    [0, 2, 2, 0],
    [0, 0, 2, 2],
    [2, 4, 4, 2],
    [2, 4, 4, 4],
];

matrix.print();
console.log("----------");
move(matrix, 'up');
matrix.print();


/**
 * 根据指定方向移动矩阵中的元素
 * @param {number[][]} matrix - 二维矩阵
 * @param {string} direction - 移动方向，可以是 'up', 'down', 'left', 'right'
 */
function move(matrix, direction) {
    const rows = matrix.length;
    const cols = matrix[0].length;

    /**
     * 检查坐标 (i, j) 是否在矩阵范围内
     * @param {number} i - 行坐标
     * @param {number} j - 列坐标
     * @returns {boolean} - 如果在范围内返回 true，否则返回 false
     */
    function _inRange(i, j) {
        return i >= 0 && i < rows && j >= 0 && j < cols;
    }

    // 定义根据方向计算下一个坐标的函数
    const next = {
        up: (i, j) => [i + 1, j],
        down: (i, j) => [i - 1, j],
        left: (i, j) => [i, j + 1],
        right: (i, j) => [i, j - 1],
    }

    /**
     * 找到在指定方向上的下一个非零值的坐标
     * @param {number} i - 起始行坐标
     * @param {number} j - 起始列坐标
     * @returns {number[]} - 包含下一个非零值的坐标和值的数组，如果没有找到则返回 undefined
     */
    function _nextNonZeroValue(i, j) {
        if (!_inRange(i, j)) {
            return;
        }

        let [nextI, nextJ] = next[direction](i, j);
        while (_inRange(nextI, nextJ)) {
            if (matrix[nextI][nextJ] !== 0) {
                return [nextI, nextJ, matrix[nextI][nextJ]];
            }

            [nextI, nextJ] = next[direction](nextI, nextJ);
        }
    }


    /**
     * 执行计算以处理矩阵中的数值移动和合并
     * 根据给定的方向，寻找下一个非零值并尝试将其移动到当前位置或与当前位置的值合并
     *
     * @param {number} i - 当前行索引
     * @param {number} j - 当前列索引
     */
    function _cal(i, j) {
        if (!_inRange(i, j)) {
            return;
        }
        const result = _nextNonZeroValue(i, j);
        if (!result) {
            return;
        }
        const [nextI, nextJ, nextValue] = result;
        if (matrix[i][j] === 0) {
            matrix[i][j] = nextValue;
            matrix[nextI][nextJ] = 0;
            _cal(i, j);
        }
        else if (matrix[i][j] === nextValue) {
            matrix[i][j] *= 2;
            matrix[nextI][nextJ] = 0;
        }

        [i, j] = next[direction](i, j);
        _cal(i, j);
    }

    // 根据移动方向处理矩阵
    if (direction === 'up') {
        for (let j = 0; j < cols; j++) {
            _cal(0, j);
        }
    }
    else if (direction === 'down') {
        for (let j = 0; j < cols; j++) {
            _cal(rows - 1, j);
        }
    }
    else if (direction === 'left') {
        for (let i = 0; i < rows; i++) { }
    }
    else if (direction === 'right') {
        for (let i = 0; i < rows; i++) {
            _cal(i, cols - 1);
        }
    }
}